docs/proposals/20240327-enhancedlivenessprobe.md

```diff
@@ -0,0 +1,301 @@
+
+
+### API Definition
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+  ...
+    // [user defined] using this enhanced livnessProbe, otherwise the native checking
+    apps.kruise.io/using-enhanced-liveness: "true"
+  ...
+  
+  labels:
+  ...
+  ...
+  spec:
+  ...
+    livenessProbe:
+      failureThreshold: 3
+      httpGet:
+        path: /health
+        port: 7002
+        scheme: HTTP
+      initialDelaySeconds: 3000
+      periodSeconds: 100
+      successThreshold: 1
+      timeoutSeconds: 5
+...
+```
+
+
+### Functional Assessment
+| enhanced(this solution)                                                                                                                                                                         | native                            | 
+|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------| 
+| ● specified component detection(nodePodProbe controller in kruise-daemonset)  + global component decision(enhancedLivenessProbeController in kruise-manager) in cluster                         | ● single machine detection + single machine decision(all in kubelet) | 
+| ● kruiseDaemon/ProbeDetector detection enabled in bypaas,  strong expansibility                                                                                                                 | ● no                              | 
+| ● the global perspective of the application, the maxUnavailable protection policy and the capactiy building of the resilience protection system(Kruise-manager/EnhancedLivenessProbeController) | ● no                              | 
+| ● the architecture is easy to expand, without invasive modification of native components, adapt easily to different scaenarios with strong scalability                                          | ● strong coupling, poor scalability | 
+
+
+### Other Notes
+● Since the probe detection is implemented by the nodePodProbe controller in Kruise-daemonset, for the pod without ownerreference, this solution can also take effect on the enhancement ability for restarting containers when in failure livenessProbe status. 
+
+● Pod workload need to use the PodUnavailableBudget in kruise, this solution should enable the PubFeature.
+
+● Override the community logic, both the native livenessProbe configuration and this solution configuration exist in the process.
+
+### Release Plan
+● NodePodProbe controller in kruise-daemonset supports TCP/HTTPGET/RPC probe besides EXEC checking.
+
+● EnhancedLivenessProbeWebhook is developed firstly to convert the standard livenessProbe configuration in the special field in pod annotations.
+
+● EnhancedLivenessProbeMapNodePodProbeController is developed to convert the probe config to the nodePodProbe custom resource in kruise.
```

Comment:
Name(EnhancedLivenessProbeMapNodePodProbeController ) is too long , consider rename it to LivenessNodeProbeController